### ISSUE TITLE
pbft: bound number of requests

### ISSUE THREAD

**Issue by corecode** [hyperledger-archives/fabric#1171]
This is for the PBFT implementation for #1098.
- Bound the number of new requests we accept from the local peer.
- Bound the number of requests any other replica proposes.
- Reject primary if it proposes too many requests from any replica.

@cca88 @vukolic @jyellick @kchristidis @tuand27613

**Comment by corecode** [comment]
This is the current status (batch):
- The number of requests we accept from the local peer is mediated by the number of requests we can send to the primary.  This fails when we are the primary or we can load the network ininitely (e.g. we are not connected at all)
- It is unclear to me how to bound the number of requests per replica reliably, because multiple pbft requests are in flight, and a new request may arrive before the old one is retired.  The watermarks do limit the number of in flight PBFT requests in any case.

**Comment by kchristidis** [comment]
Based on our conversations today, this is how I'm parsing the current thinking regarding locally generated requests:

We will need two buffered channels. One that acts as a queue for the requests generated locally before we start dropping any additional arrivals (or before we block the process that attempts to inject any new requests into the system). Another one that that signals the number of locally generated requests that are currently being processed by the PBFT goroutine.

An `engine` goroutine ranges on the first channel, and routes incoming transactions to `RecvRequest` in `external.go`. `RecvRequest` will not inject the transaction into the PBFT manager's queue until it's able to write to the second channel.

Injecting the transaction means assigning it to one of the empty slots that the replica holds for its locally generated transactions. (The size of the second buffered channel is equal to the number of slots that the replica holds for these transactions, and the number of empty slots in the channel is equal to the number of empty slots that the replica holds in `outstandingRequests` for its own locally generated transactions.)

Whenever a transaction is executed, the replica removes it from the "local" slots it keeps in `outstandingRequests` and it reads from the second channel to signal that a new locally generated transaction can be consumed by PBFT.

There is no need for `RecvRequest` to know which local slot number are available, etc. The method can be agnostic to them, so the second channel can just be a `chan struct{}` (instead of a `chan int`).

This implementation should bound the number of new requests we accept from the local peer. (Item 1 in this issue.)

Let me know if I am missing something, or post in Slack.

**Comment by jyellick** [comment]
Although your description might work, at least in my head I think it's more natural to flip the direction of the second channel.

In paragraph 3, say "... will not inject the transactions into the PBFT manager's queue until it's able to _read_ from the second channel."

In paragraph 4, say "... and the number of empty slots in the channel is equal to _the number of assigned slots_ that the replica holds ..."

In this way, if we chose to make the optimization of a `chan int`, it would be the PBFT side supplying the slot, rather than the other way around.

**Comment by kchristidis** [comment]
You're right — flipping the direction of the second channel makes more sense.

**Comment by corecode** [comment]
I think that's what I already have in code.
